File: bitcoin-stockmem-ts/src/evaluation/metrics.ts
/**
 * Evaluation metrics: Accuracy and MCC.
 */

export function accuracy(predictions: string[], actuals: string[]): number {
  if (!predictions.length) return 0;
  const correct = predictions.filter((p, i) => p === actuals[i]).length;
  return correct / predictions.length;
}

export function mcc(predictions: string[], actuals: string[]): number {
  let tp = 0, fp = 0, tn = 0, fn = 0;
  for (let i = 0; i < predictions.length; i++) {
    if (predictions[i] === "up" && actuals[i] === "up") tp++;
    else if (predictions[i] === "up" && actuals[i] === "down") fp++;
    else if (predictions[i] === "down" && actuals[i] === "down") tn++;
    else if (predictions[i] === "down" && actuals[i] === "up") fn++;
  }
  const denom = Math.sqrt((tp + fp) * (tp + fn) * (tn + fp) * (tn + fn));
  return denom === 0 ? 0 : (tp * tn - fp * fn) / denom;
}

export function evaluate(predictions: string[], actuals: string[]) {
  return {
    accuracy: Math.round(accuracy(predictions, actuals) * 10000) / 10000,
    mcc: Math.round(mcc(predictions, actuals) * 10000) / 10000,
    total: predictions.length,
    correct: predictions.filter((p, i) => p === actuals[i]).length,
  };
}

File: bitcoin_stockmem/evaluation/metrics.py
"""Evaluation metrics: Accuracy and Matthews Correlation Coefficient."""

from __future__ import annotations

import math


def accuracy(predictions: list[str], actuals: list[str]) -> float:
    """Standard accuracy: correct / total."""
    if not predictions:
        return 0.0
    correct = sum(1 for p, a in zip(predictions, actuals) if p == a)
    return correct / len(predictions)


def mcc(predictions: list[str], actuals: list[str]) -> float:
    """Matthews Correlation Coefficient for binary classification.

    Treats "up" as positive and "down" as negative.
    MCC = (TP*TN - FP*FN) / sqrt((TP+FP)*(TP+FN)*(TN+FP)*(TN+FN))
    """
    tp = fp = tn = fn = 0
    for pred, actual in zip(predictions, actuals):
        if pred == "up" and actual == "up":
            tp += 1
        elif pred == "up" and actual == "down":
            fp += 1
        elif pred == "down" and actual == "down":
            tn += 1
        elif pred == "down" and actual == "up":
            fn += 1

    denom = math.sqrt(
        (tp + fp) * (tp + fn) * (tn + fp) * (tn + fn)
    )
    if denom == 0:
        return 0.0
    return (tp * tn - fp * fn) / denom


def evaluate(predictions: list[str], actuals: list[str]) -> dict:
    """Compute all metrics and return as a dict."""
    return {
        "accuracy": round(accuracy(predictions, actuals), 4),
        "mcc": round(mcc(predictions, actuals), 4),
        "total": len(predictions),
        "correct": sum(1 for p, a in zip(predictions, actuals) if p == a),
    }
